### ISSUE TITLE
Vocoders performance / universality

### ISSUE THREAD

**Issue by ghost** [TensorSpeech/TensorFlowTTS#565]
Hello !
I just rewrote the WaveGlow vocoder in tensorflow trying to improve its performance but it fails (1.7 sec generated / sec with the pytorch model drops down to less than 1.2 generated / sec in tensorflow)

What are the performances of your vocoders ? are they better ? Which kind of performance do you achieve for a full inference (text -> mel -> audio) with your vocoders ? And which one is the fastest ?

Also are they « universal » ? By universal I mean they can be used to transform **any** spectrogram (from any speaker in any language) to the corresponding audio ? The WaveGlow vocoder I use is universal but relatively slow and I would like to change in order to improve performances of my tts pipeline

Thank you in advance !

yui 😋 

ps : the WaveGlow model is entirely copied from the NVIDIA’s pretrained model available on torch hub and the 2 performances listed above are for complete inference (with my implementation of `tacotron2` as synthesizer)

**Comment by dathudeptrai** [comment]
@ZDisket can you help him ?

**Comment by ZDisket** [comment]
@yui-dl Most vocoders can be universal if trained on a lot of speakers from various datasets. Even one MelGAN-STFT model I trained on 20 speakers a while ago kind of worked as a universal vocoder, but not that well; you also probably want to increase the size of the model as shown in the Universal MelGAN paper: https://arxiv.org/abs/2011.09631. Also, MelGAN and especially multi-band are very fast.

For myself, I abandoned universal vocoders in favor of adapting pretrained vocoder model to new speakers, as the quality and speech characteristics loss was unacceptable, and I had found that after a learning rate reduction, it's possible to fine-tune a Multi-Band MelGAN with HiFi-GAN discriminator (MPD + MRD) model to adapt to a new speaker pretty well within 30k steps. I still need to test it more.

>  Which kind of performance do you achieve for a full inference (text -> mel -> audio) with your vocoders ?

Using (larger, 512 filters) mb-melgan, I get about 0.360 RTF with Tacotron 2, and 0.078 with FastSpeech2.

**Comment by yui-dl** [comment]
Thank you for your reply !
Quite impressive results in performance but I need a universal vocoder because I try to improve my `SV2TTSTacotron2` model, which aims to be a universal synthesizer (https://arxiv.org/abs/1806.04558) (so need of a universal vocoder :smile: )

Note that if you are interested in, the [NVIDIA WaveGlow vocoder](https://github.com/NVIDIA/waveglow) is really good vocoder (but relatively slow) and I just added a tensorflow implementation on [my TTS repository](https://github.com/yui-dl/text_to_speech)

**Comment by Freddy-pp** [comment]
> Using (larger, 512 filters) mb-melgan, I get about 0.360 RTF with Tacotron 2, and 0.078 with FastSpeech2 on a Ryzen 5 3600.

Hello, @ZDisket. Does increasing number of filters affects quality? And what else should be changed besides the number of filters to improve the quality?

**Comment by ZDisket** [comment]
@Freddy-pp I can't confirm because aside from that change I added the MPD and while I did see a noticeable increase in multispeaker performance, I don't know which change contributed more.
My theory is that larger model = better at multi-speaker.

**Comment by ZDisket** [comment]
@yui-dl 
>Quite impressive results in performance but I need a universal vocoder 

As I stated before, most if not all vocoders on this repo can be universal with the right tuning. I think your best bet will be large Multi-Band MelGAN with HiFi-GAN discriminator. I can give you a training script and config as I haven't PRed it yet.

**Comment by yui-dl** [comment]
Yes indeed sorry my bad
What is the training time of these models ? Because I just have 1 GPU with 6Go RAM so  it’s more a power limitation in my case and I am not sure to be able to train it enough to have comparable results (in term of quality) compared to my current vocoder 😄 
If I have the time this summer and if it is okay for you I would be quite interested to integrate some of your vocoders to my project ! (I always prefer copy models / codes to better understand them instead of just running an existing project 😄 )

Note that if you want I have pretrained Tacotron2 in Fr if you want to add it in your supported languages 😃  
The `Tacotron2` class is basically a mixte of your implementation and the NVIDIA one (with extension for `SV2TTS` support) so you should be able to integrate it relatively easily

**Comment by tts-nlp** [comment]
hi @dathudeptrai @ZDisket , when using multi speaker data to training mb melgan , the loss explode, how to solves this! thanks
![Uploading Screenshot from 2021-05-19 17-36-29.png…]()

**Comment by tts-nlp** [comment]
hi @dathudeptrai @ZDisket , when using multi speaker data to training mb melgan , the loss explode, how to solves this! thanks
![Screenshot from 2021-05-19 17-36-29](https://user-images.githubusercontent.com/58066364/118792370-2f377400-b8ca-11eb-84f3-87fffc8319dc.png)

**Comment by dathudeptrai** [comment]
@tts-nlp https://github.com/TensorSpeech/TensorFlowTTS/issues/247

**Comment by tts-nlp** [comment]
@dathudeptrai thanks! , I will try this!

**Comment by ZDisket** [comment]
@yui-dl 
>What is the training time of these models ? 


About 2 days for stock MB-MelGAN on a Tesla V100, twice that. For the memory limitation, you'll probably need to use gradient accumulation.

>If I have the time this summer and if it is okay for you I would be quite interested to integrate some of your vocoders to my project ! 

You'll have to ask @dathudeptrai

**Comment by Freddy-pp** [comment]
> @Freddy-pp I can't confirm because aside from that change I added the MPD and while I did see a noticeable increase in multispeaker performance, I don't know which change contributed more.
> My theory is that larger model = better at multi-speaker.

Thank you.

> I think your best bet will be large Multi-Band MelGAN with HiFi-GAN discriminator. I can give you a training script and config as I haven't PRed it yet.

I'm also interested in a large Multi-Band MelGAN with HiFi-GAN discriminator training script and config if you could share your code. Thanks in advance.

**Comment by ZDisket** [comment]
@Freddy-pp It's in the .zip. Note that to use it as-is you need to resample your files to 44.1KHz, so I've attached a preprocessing config as well.
[mbmelganhfd.zip](https://github.com/TensorSpeech/TensorFlowTTS/files/6548464/mbmelganhfd.zip)

**Comment by Freddy-pp** [comment]
Thanks a lot. :)

**Comment by yingfenging** [comment]
@ZDisket  hello , 

> For myself, I abandoned universal vocoders in favor of adapting pretrained vocoder model to new speakers, as the quality and speech characteristics loss was unacceptable, and I had found that after a learning rate reduction, it's possible to fine-tune a Multi-Band MelGAN with HiFi-GAN discriminator (MPD + MRD) model to adapt to a new speaker pretty well within 30k steps. I still need to test it more. I'm aiming for faster than realtime on average consumer PCs, so I haven't had any experience with anything other than GANs for vocoders.

The "Multi-Band MelGAN with HiFi-GAN discriminator (MPD + MRD) " means : use the MPD(multi-period waveform discriminator from HiFi-GAN) and MRD (multi-resolution spectrogram discriminator which come from universal melgan) ?

**Comment by ZDisket** [comment]
@yingfenging Almost yes, the MRD refers to the multi-resolution discriminator already found in MB-MelGAN.

**Comment by yingfenging** [comment]
@ZDisket  ok, i want to confirm again. The MRD refers the blow. Thanks in advance.
![image](https://user-images.githubusercontent.com/49307961/124243146-13bdbb00-db50-11eb-9639-38ea298b711b.png)

**Comment by ZDisket** [comment]
@yingfenging I don't know which paper you got that graphic from but that seems to be it, yes.

**Comment by yingfenging** [comment]
> @yingfenging I don't know which paper you got that graphic from but that seems to be it, yes.

hello, It come from  the paper MelGAN ( https://arxiv.org/abs/1910.06711 ) , the 4th page.

**Comment by GuangChen2016** [comment]
@ZDisket 
Hi, I also tested MPD for fullband melgan but found some distinct trills and jitters in the synthesized audios, did it appear in your experiments? In details, the mel-spectrum in corresponding ares discontinuous, shown as follows:
![LEEGHJGKZO(@0@F51_LO$XL](https://user-images.githubusercontent.com/23181283/159401940-825c8dd0-2bd5-4471-97de-9649f605fc14.png)
Do you have any ideas to solve this? Thanks.
